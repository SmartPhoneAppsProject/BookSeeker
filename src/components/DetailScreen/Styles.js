import { StyleSheet } from 'react-native';

export const index = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  body: {
    flex: 1,
  },
});

export const bookImage = StyleSheet.create({
  image: {
    flex: 1,
    width: undefined,
    height: undefined,
    marginTop: 10,
  },
});

export const tagsList = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
  },
  scrollView: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  tag: {
    margin: 3,
  },
  tagText: {
    paddingRight: 5,
    color: '#808080',
  },
});

export const bookInfo = StyleSheet.create({
  container: {
    flex: 1.5,
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  title: {
    flex: 1,
    fontSize: 20,
  },
  date: {
    flex: 1,
    fontSize: 14,
    color: '#6E6E6E',
  },
});

export const borrowReturnButton = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  title: {
    fontWeight: '700',
  },
  button: {
    width: 100,
    height: 60,
  },
  returnColor: {
    backgroundColor: '#cd5c5c',
  },
  borrowColor: {
    backgroundColor: '#2e8b57',
  },
});
