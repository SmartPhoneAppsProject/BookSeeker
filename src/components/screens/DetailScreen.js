import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  ScrollView,
  Image,
  Dimensions,
} from 'react-native';

import {
  Button,
} from 'react-native-elements';

import {
  MaterialCommunityIcons,
  MaterialIcons,
  Octicons
} from '@expo/vector-icons';

import { icon } from "../../modules/Icons";
import { rentBook } from '../../modules/Network';

export default class DetailScreen extends Component {
  static navigationOptions = {
    title: 'DetailScreen',
  };

  constructor(props) {
    super(props);

    this.state = {
      currentStatus: this.props.navigation.state.params.item.status
    };

    this._lendBook = this._lendBook.bind(this);
    this._returnBook = this._returnBook.bind(this);
  }

  _lendBook = () => {
    this.setState({
      currentStatus: true
    });

    const json = JSON.stringify({
      jan_code: this.props.navigation.state.params.item.jan_code,
      status: true
    });

    this.changeBookStatus(json);
  };

  _returnBook = () => {
    this.setState({
      currentStatus: false
    });

    const json = JSON.stringify({
      jan_code: this.props.navigation.state.params.item.jan_code,
      status: false
    });

    this.changeBookStatus(json);
  };

  changeBookStatus = (json) => {
    rentBook(json)
      .then(response => response.json())
      .then(responseJson => {
        console.log(responseJson);
      })
      .catch(error => {
        console.warn(error);
        rentBook(json)
          .then(response => response.json())
          .then(responseJson => {
            console.log(responseJson);
          })
          .catch(error => console.error(error));
      });
  };

  _renderImage = () => {
    let imageUri = this.props.navigation.state.params.item.image;
    if (imageUri === 'none') {
      imageUri = 'https://facebook.github.io/react/logo-og.png';
    }

    return (
      <View style={[styles.base, styles.imgContainer]}>
        <Image style={styles.img} source={{ uri: imageUri }}/>
      </View>
    )
  };

  _renderTags = () => {
    const tags = this.props.navigation.state.params.item.tags;

    let formated = [];
    let tag;
    for (tag of tags) {
      tag = <Text style={styles.tagText}>{icon(tag.name)}{tag.name}</Text>;
      formated.push(
        <View style={styles.tag} key={tag.key}>
          {tag}
        </View>
      )
    }

    return (
      <View style={[styles.base, styles.tagContainer]}>
        <ScrollView horizontal={true}>
          {formated}
        </ScrollView>
      </View>
    )
  };

  _renderIcon = () => {
    if (this.state.currentStatus === true) {
      return (
        <View style={[styles.base, styles.status]}>
          <Octicons name='circle-slash' size={40} color='#cd5c5c'/>
        </View>
      )
    } else {
      return (
        <View style={[styles.base, styles.status]}>
          <MaterialCommunityIcons name='check-circle-outline' size={40} color='#2e8b57'/>
        </View>
      )
    }
  };

  _renderButton = () => {
    if (this.state.currentStatus === true) {
      return (
        <View style={[styles.base, styles.buttonContainer]}>
          <Button
            icon={<MaterialCommunityIcons name='keyboard-return' size={30} color='white'/>}
            title="返却"
            titleStyle={{ fontWeight: "700" }}
            buttonStyle={{ width: 100, height: 60, backgroundColor: '#cd5c5c' }}
            iconContainerStyle={{ marginRight: 10 }}
            onPress={this._returnBook}/>
        </View>
      )
    } else {
      return (
        <View style={[styles.base, styles.buttonContainer]}>
          <Button
            icon={<MaterialCommunityIcons name='book-open-page-variant' size={30} color='white'/>}
            title="貸出"
            titleStyle={{ fontWeight: "700" }}
            buttonStyle={{ width: 100, height: 60, backgroundColor: '#2e8b57' }}
            iconContainerStyle={{ marginRight: 10 }}
            onPress={this._lendBook}
          />
        </View>
      )
    }
  };

  render() {
    const { params } = this.props.navigation.state;
    const renderImage = this._renderImage();
    const renderTags = this._renderTags();
    const renderIcon = this._renderIcon();
    const renderButton = this._renderButton();

    return (
      <View style={styles.container}>
        {renderImage}
        <View style={styles.mainContainer}>
          {renderTags}
          <View style={styles.titleContainer}>
            <Text style={styles.title}>{params.item.title}</Text>
            <Text style={styles.date}>出版日：{params.item.published_at}</Text>
          </View>
          <View style={styles.statusContainer}>
            {renderIcon}
            {renderButton}
          </View>
        </View>
      </View>
    );
  }
}

const { width } = Dimensions.get('window');

const styles = StyleSheet.create({
  container:
    {
      flex: 1,
      backgroundColor: '#fff',
    },
  base:
    {
      justifyContent: 'center',
      alignItems: 'center',
    },
  imgContainer:
    {
      flex: 1,
    },
  img:
    {
      height: 300,
      width: width,
    },
  mainContainer:
    {
      flex: 1
    },
  tagContainer:
    {
      flex: 0.5,
      margin: 10,
    },
  tag:
    {
      margin: 3,
      backgroundColor: '#f5f5f5',
      borderColor: '#f5f5f5',
      borderRadius: 4,
      borderWidth: 1,
      overflow: 'hidden',
      paddingRight: 8,
      paddingLeft: 8,
      paddingTop: 13,
    },
  tagText:
    {
      paddingRight: 5,
      color: '#808080',
    },
  titleContainer:
    {
      flex: 1,
      justifyContent: 'center',
      marginLeft: 50,
    },
  title:
    {
      fontSize: 20,
      marginBottom: 10,
    },
  dateContainer:
    {
      flex: 0.4,
    },
  date:
    {
      fontSize: 14,
      color: '#6E6E6E',
    },
  statusContainer:
    {
      flex: 1,
      flexDirection: 'row',
      marginBottom: 15,
    },
  status:
    {
      flex: 1,
    },
  buttonContainer:
    {
      flex: 1,
    },
  button:
    {
      width: 70,
    },
});
