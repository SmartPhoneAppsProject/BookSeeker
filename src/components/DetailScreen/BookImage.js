import React from 'react';
import PropTypes from 'prop-types';
import {
  View,
  Image,
} from 'react-native';

import { bookImage as styles } from './Styles';

const NoImage = () => (
  <Image
    style={styles.img}
    source={require('./../../../assets/no_image.png')}
  />
);

const NetworkImage = uri => (
  <Image
    style={styles.img}
    source={{ uri }}
  />
);

const BookImage = ({ uri }) => (
  <View style={[styles.base, styles.imgContainer]}>
    {uri ?
      <NetworkImage uri={uri} />
      : <NoImage />
    }
  </View>
);

BookImage.propTypes = {
  uri: PropTypes.string.isRequired,
};

export default BookImage;
