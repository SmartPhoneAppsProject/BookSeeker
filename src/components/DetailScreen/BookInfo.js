import React from 'react';
import PropTypes from 'prop-types';
import {
  View,
  Text,
} from 'react-native';

import { bookInfo as styles } from './Styles';

const BookInfo = ({ title, published }) => (
  <View style={styles.container}>
    <Text style={styles.title}>{title}</Text>
    <Text style={styles.date}>出版日：{published}</Text>
  </View>
);

BookInfo.propTypes = {
  title: PropTypes.string.isRequired,
  published: PropTypes.string.isRequired,
};

export default BookInfo;
