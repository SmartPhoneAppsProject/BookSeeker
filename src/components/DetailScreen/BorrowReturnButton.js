import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import { Button } from 'react-native-elements';
import {
  Octicons,
  MaterialCommunityIcons,
} from '@expo/vector-icons';

import { borrowReturnButton as styles } from './Styles';

const ReturnButton = ({ status, navigate }) => (
  <Button
    icon={
      <Octicons
        name="circle-slash"
        size={30}
        color="#ffffff"
      />
    }
    title="返却"
    titleStyle={styles.title}
    buttonStyle={[styles.button, styles.returnColor]}
    onPress={() => navigate('LentScan', { status })}
  />
);

const BorrowButton = ({ status, navigate }) => (
  <Button
    icon={
      <MaterialCommunityIcons
        name="check-circle-outline"
        size={30}
        color="#ffffff"
      />
    }
    title="貸出"
    titleStyle={styles.title}
    buttonStyle={[styles.button, styles.borrowColor]}
    onPress={() => navigate('LentScan', { status })}
  />
);

const BorrowReturnButton = ({ status, navigate }) => (
  <View style={styles.container}>
    {status ?
      <ReturnButton status={status} navigate={navigate} />
      : <BorrowButton status={status} navigate={navigate} />
    }
  </View>
);

BorrowReturnButton.propTyles = {
  status: PropTypes.bool.isRequired,
  navigate: PropTypes.object.isRequired,
};

export default BorrowReturnButton;
