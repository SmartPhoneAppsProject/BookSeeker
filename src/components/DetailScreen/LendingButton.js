import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import { Button } from 'react-native-elements';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import { lendingButton as styles } from './Styles';

const LendingButton = ({ status, navigate }) => {
  if (status === true) {
    return (
      <View style={[styles.base, styles.buttonContainer]}>
        <Button
          icon={<MaterialCommunityIcons name="keyboard-return" size={30} color="white" />}
          title="返却"
          titleStyle={{ fontWeight: '700' }}
          buttonStyle={{ width: 100, height: 60, backgroundColor: '#cd5c5c' }}
          iconContainerStyle={{ marginRight: 10 }}
          onPress={() => navigate('LentScan', { action: 'return' })}
        />
      </View>
    );
  }

  return (
    <View style={[styles.base, styles.buttonContainer]}>
      <Button
        icon={<MaterialCommunityIcons name="book-open-page-variant" size={30} color="white" />}
        title="貸出"
        titleStyle={{ fontWeight: '700' }}
        buttonStyle={{ width: 100, height: 60, backgroundColor: '#2e8b57' }}
        iconContainerStyle={{ marginRight: 10 }}
        onPress={() => navigate('LentScan', { action: 'lend' })}
      />
    </View>
  );
};

LendingButton.propTyles = {
  status: PropTypes.bool.isRequired,
  navigate: PropTypes.object.isRequired,
};

export default LendingButton;
