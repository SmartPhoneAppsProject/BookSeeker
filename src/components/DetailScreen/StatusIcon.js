import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import {
  Octicons,
  MaterialCommunityIcons,
} from '@expo/vector-icons';

import { statusIcon as styles } from './Styles';

const LendIcon = () => (
  <View style={[styles.base, styles.status]}>
    <Octicons name="circle-slash" size={40} color="#cd5c5c" />
  </View>
);

const BorrowIcon = () => (
  <View style={[styles.base, styles.status]}>
    <MaterialCommunityIcons name="check-circle-outline" size={40} color="#2e8b57" />
  </View>
);

const StatusIcon = ({ status }) => {
  if (status === true) {
    return <LendIcon />;
  }

  return <BorrowIcon />;
};

StatusIcon.propTyles = {
  status: PropTypes.bool.isRequired,
};

export default StatusIcon;
