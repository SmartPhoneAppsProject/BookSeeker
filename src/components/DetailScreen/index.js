import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';

import { index as styles } from './Styles';
import BookImage from './BookImage';
import TagsList from './TagsList';
import BookInfo from './BookInfo';
import BorrowReturnButton from './BorrowReturnButton';

class DetailScreen extends Component {
  render() {
    const { navigate } = this.props.navigation;
    const {
      tags,
      status,
      image,
      title,
      published,
    } = this.props.book;

    return (
      <View style={styles.container}>
        <BookImage uri={image} />
        <View style={styles.body}>
          <TagsList tags={tags} />
          <BookInfo
            title={title}
            published={published}
            tags={tags}
          />
          <BorrowReturnButton
            status={status}
            navigate={navigate}
          />
        </View>
      </View>
    );
  }
}

DetailScreen.navigationOptions = {
  title: '詳細',
};

DetailScreen.propTypes = {
  book: PropTypes.shape({
    tags: PropTypes.array.isRequired,
    status: PropTypes.bool.isRequired,
    image: PropTypes.string,
    title: PropTypes.string.isRequired,
    published: PropTypes.string.isRequired,
  }).isRequired,
};

export default DetailScreen;
