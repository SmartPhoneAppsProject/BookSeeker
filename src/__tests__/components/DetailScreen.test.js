import React from 'react';
import { shallow } from 'enzyme';
import DetailScreen from '../../components/DetailScreen';
import _books from '../../api/books.json';

describe('<EntryScreen />', () => {
  it('正しくレンダリングされていること', () => {
    const navigation = {
      state: {
        params: {
          item: _books[0],
        },
      },
    };
    const wrapper = shallow(<DetailScreen
      book={_books[0]}
      navigation={navigation}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
