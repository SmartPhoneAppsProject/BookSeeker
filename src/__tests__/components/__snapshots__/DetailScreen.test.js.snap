// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailScreen /> 正しくレンダリングされていること 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DetailScreen
    book={
      Object {
        "created_at": "2018-03-09T00:00:00Z",
        "id": 2,
        "image": "none",
        "isbn": "9784774184111",
        "published": "2018-02-09T00:00:00Z",
        "status": true,
        "tags": Array [],
        "title": "javascript ",
        "updated_at": "2018-03-09T00:00:00Z",
      }
    }
    navigation={
      Object {
        "state": Object {
          "params": Object {
            "item": Object {
              "created_at": "2018-03-09T00:00:00Z",
              "id": 2,
              "image": "none",
              "isbn": "9784774184111",
              "published": "2018-02-09T00:00:00Z",
              "status": true,
              "tags": Array [],
              "title": "javascript ",
              "updated_at": "2018-03-09T00:00:00Z",
            },
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <BookImage
          uri="none"
        />,
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <TagsList
            tags={Array []}
          />
          <BookInfo
            published="2018-02-09T00:00:00Z"
            tags={Array []}
            title="javascript "
          />
          <BorrowReturnButton
            navigate={undefined}
            status={true}
          />
        </View>,
      ],
      "style": Object {
        "backgroundColor": "#fff",
        "flex": 1,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "uri": "none",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <TagsList
              tags={Array []}
            />,
            <BookInfo
              published="2018-02-09T00:00:00Z"
              tags={Array []}
              title="javascript "
            />,
            <BorrowReturnButton
              navigate={undefined}
              status={true}
            />,
          ],
          "style": Object {
            "flex": 1,
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "tags": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "published": "2018-02-09T00:00:00Z",
              "tags": Array [],
              "title": "javascript ",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "navigate": undefined,
              "status": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <BookImage
            uri="none"
          />,
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <TagsList
              tags={Array []}
            />
            <BookInfo
              published="2018-02-09T00:00:00Z"
              tags={Array []}
              title="javascript "
            />
            <BorrowReturnButton
              navigate={undefined}
              status={true}
            />
          </View>,
        ],
        "style": Object {
          "backgroundColor": "#fff",
          "flex": 1,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "uri": "none",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <TagsList
                tags={Array []}
              />,
              <BookInfo
                published="2018-02-09T00:00:00Z"
                tags={Array []}
                title="javascript "
              />,
              <BorrowReturnButton
                navigate={undefined}
                status={true}
              />,
            ],
            "style": Object {
              "flex": 1,
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "tags": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "published": "2018-02-09T00:00:00Z",
                "tags": Array [],
                "title": "javascript ",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "navigate": undefined,
                "status": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
